security/s2a/internal/crypter/expander.go

```diff
@@ -12,3 +14,24 @@ type hkdfExpander interface {
 	// in number of bytes, and returns the resulting expanded key.
 	expand(hash func() hash.Hash, secret, label []byte, length int) ([]byte, error)
 }
+
+// defaultHKDFExpander is the default HKDF expander which uses Go's crypto/hkdf
+// for HKDF expansion.
+type defaultHKDFExpander struct{}
+
+// newDefaultHKDFExpander creates an instance of the default HKDF expander.
+func newDefaultHKDFExpander() hkdfExpander {
+	return &defaultHKDFExpander{}
+}
+
+func (*defaultHKDFExpander) expand(hash func() hash.Hash, secret, label []byte, length int) ([]byte, error) {
```

Comment:
Thinking about this API again. Is the output length related to the returned hash function resulting from calling `hash`? If that's the case, can we get the `length` from that?

Reply:
I believe we need to pass in `length` for efficiency sake. `hkdf.Expand` returns an `io.Reader` interface which exposes only a `Read(p []byte) (n int, err error)` method. We need to pass in a correctly allocated buffer here so that we can do 1 buffer allocation and 1 read rather than numerous small reads. I don't believe it's possible to get the `length` directly from the resulting call of `hkdf.Expand`. If we want to get the size directly from the hash function, we need to do `expander := hmac.New(hash, secret).Size()` before we allocate the buffer and call `hkdf.Expand`, which seems inefficient. I think it's best to leave the `length` input as it is. Please correct if I'm wrong!

Reply:
Calling `hash` returns a `hash.Hash` object [1] which has a `Size` API. There's no need to create an HMAC object to get the size of the underlying hash function. You can simply do:

```
length := hash().Size()
```

[1] https://golang.org/pkg/hash/#Hash

Reply:
Also, maybe we should rename the `hash` parameter to `h` to differentiate that from the imported `hash` package. We should do the same in the interface as well. Feel free to do that in this PR.

Reply:
Done.